### PR TITLE
Created PS script for evaluating MSBuild properties

### PR DESCRIPTION
**Pull request by jonthysell**
We often have need of determining a MSBuild property's value for a given project, but reading the raw XML files means we don't evaluate variables or conditions.

This script loads `Microsoft.Build.dll` and uses the build engine to evaluate the value of a property (without actually running the build).

Example usage:

```
PS C:\code\rnw\packages\playground> ..\..\vnext\Scripts\Eval-MsBuildProperty.ps1 -SolutionFile .\windows\playground-win32.sln -ProjectFile .\windows\playground-win32\Playground-win32.vcxproj -PropertyName WinUI2xVersion
2.6.1-prerelease.210709001
PS C:\code\rnw\packages\playground> ..\..\vnext\Scripts\Eval-MsBuildProperty.ps1 -SolutionFile .\windows\playground.sln -ProjectFile .\windows\playground\Playground.vcxproj -PropertyName WinUI2xVersion
2.6.0
```

The `ExperimentalFeatures.props` file used by the playground apps (see issue #8591) conditionally specifies different versions of `WinUI2xVersion` depending on which of the playground solutions is being used, so we can see that the script find the correct version without manually parsing `ExperimentalFeatures.props`.

Note that while this script should prove useful in a variety of situations, it is a relatively expensive operation and should only be used when absolutely necessary.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/8734)

### CODE REVIEW

**Comment by NickGerleman** [comment]
Ooh, this also would help us fix deploy when AppX signing is disabled.

**Comment by NickGerleman** [comment]
Do you have a sense of how expensive an invocation is? I'm wondering if we would end up repeating any expensive work if we called this script multiple times.

**Comment by jonthysell** [comment]
> Do you have a sense of how expensive an invocation is? I'm wondering if we would end up repeating any expensive work if we called this script multiple times.

With the examples I gave above, it's about ~800-1000ms for resolving WinUI2xVersion for playground. The longest parts are:
1. Loading the DLL into powershell
2. Processing the solution and project with MSBuild
3. Looking up the target property and evaluating it

Each one deserves to be looked at and can possibly be optimized.

The other thing we could do is, for step 3, evaluate multiple (all?) properties at once, and maybe cache the result.

File: packages/playground/windows/Playground-win32 (Package)/Playground-win32-packaging.proj [hunk 1]
```diff
@@ -64,5 +64,5 @@
   </PropertyGroup>
   <Import Project="..\..\..\..\vnext\PropertySheets\WinUI.props" />
   <Import Project="$(WapProjPath)\Microsoft.DesktopBridge.targets" />
-  <Import Project="..\packages\$(WinUIPackageName).$(WinUI3Version)\build\Microsoft.WinUI.AppX.targets" />
+  <Import Project="..\packages\$(WinUIPackageName).$(WinUI3Version)\build\Microsoft.WinUI.AppX.targets" Condition="Exists('..\packages\$(WinUIPackageName).$(WinUI3Version)\build\Microsoft.WinUI.AppX.targets')" />
```
Comment (jonthysell) [inline packages/playground/windows/Playground-win32 (Package)/Playground-win32-packaging.proj:67]: This is just a bug we have in playground I hit when creating and using this script, as the file doesn't exist before a nuget restore, which my script doesn't "technically" require.

File: vnext/Scripts/msbuildeval.ps1 [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    # Look for the specified PropertyName and evaluate it
+    $project.Properties | ForEach-Object -Process {
+        if ($_.Name -eq $PropertyName) {
+            Write-Host $_.EvaluatedValue
+            break
+        }
+    }
+}
+
+$SolutionPath = [System.IO.Path]::GetFullPath((Join-Path $pwd $SolutionFile))
```
Comment (JunielKatarn) [inline vnext/Scripts/msbuildeval.ps1:70]: Why not just make `SolutionPath` the input argument?

File: vnext/Scripts/msbuildeval.ps1 [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    & $MsBuildPath\MSBuild.exe $SolutionPath | Out-Null
+
+    # Build a local project collection
+    $projectCollection =[Microsoft.Build.Evaluation.ProjectCollection]::new()
+
+    # Process solution
+    $solution = [Microsoft.Build.Evaluation.Project]::new("$SolutionPath.metaproj", $null, "Current", $projectCollection)
+
+    # Clean up "metaproj" files
+    Remove-Item "$SolutionPath.metaproj" | Out-Null
```
Comment (JunielKatarn) [inline vnext/Scripts/msbuildeval.ps1:47]: Why is this needed?

File: vnext/Scripts/msbuildeval.ps1 [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    # Build a temporary "metaproj" of the solution file so it can be processed
+    ${env:MSBUILDEMITSOLUTION} = 1
+    & $MsBuildPath\MSBuild.exe $SolutionPath | Out-Null
+
+    # Build a local project collection
+    $projectCollection =[Microsoft.Build.Evaluation.ProjectCollection]::new()
+
+    # Process solution
+    $solution = [Microsoft.Build.Evaluation.Project]::new("$SolutionPath.metaproj", $null, "Current", $projectCollection)
```
Comment (JunielKatarn) [inline vnext/Scripts/msbuildeval.ps1:44]: What exactly does `Process solution` mean?
Does it run the default (usually `Build`) target?
If so, why deal with `.metaproj`?

**Review by JunielKatarn — COMMENTED** [review]
Additional comment:
- (nit) Can we adhere to PowerShell naming conventions (`<Verb>-<Object>`) for the scripts also? (i.e. `Eval-MsBuildProperty.ps1`
- Maybe this can be achieved just wrapping around `MSBuild.exe`?\
  See `Vars` target at https://github.com/microsoft/react-native-windows/blob/main/vnext/Directory.Build.targets#L8

**Comment by dannyvv** [comment]
Just sharing as FYI: Even if you evaluate all properties, it might not be the same values as what is seen during the build. Because in MsBuild the model is that all state (property groups and item groups) are globally mutable values. So any target can set any value to any property or itemgroup.

As said, just sharing as FYI: This is way more reliable than xml parsing...

**Comment by dannyvv** [comment]
> * Maybe this can be achieved just wrapping around `MSBuild.exe`?
>   See `Vars` target at https://github.com/microsoft/react-native-windows/blob/main/vnext/Directory.Build.targets#L8

You have a cool idea here: We could have a target: `<Target Name="ExtractState">` that would write out a json file that we can then parse from js-side.

File: vnext/Scripts/msbuildeval.ps1 [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+function Get-MSBuildPath {
+    $vsWhere = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe"
+    if (!(Test-Path $vsWhere)) {
+        throw "Unable to find vswhere.exe."
+    }
+    $vsPath = & $vsWhere -version 16.5 -property installationPath;
+    return "$vsPath\MSBuild\Current\Bin\";
+}
```
Comment (NickGerleman) [inline vnext/Scripts/msbuildeval.ps1:18]: The CLI has some extra logic compared to this to determine which copy of MSBuild to use. E.g. it will only consider installations with certain packages, and has special handling of prereleases.

Consider allowing the CLI to explicitly pass its own resolved MSBuild path instead of trying to replicate the logic.

File: vnext/Scripts/msbuildeval.ps1 (hunk 3 above)
Comment (NickGerleman) [reply vnext/Scripts/msbuildeval.ps1:47]: Consider wrapping this in try/finally style block so that we still do it if the above line fails.

File: vnext/Scripts/msbuildeval.ps1 [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    # Process the project file (with the Solution* properties we calculated before)
+    $project = [Microsoft.Build.Evaluation.Project]::new("$ProjectPath", $globalProps, "Current", $projectCollection)
+
+    # Look for the specified PropertyName and evaluate it
+    $project.Properties | ForEach-Object -Process {
+        if ($_.Name -eq $PropertyName) {
+            Write-Host $_.EvaluatedValue
+            break
+        }
+    }
```
Comment (NickGerleman) [inline vnext/Scripts/msbuildeval.ps1:67]: Re multiple invocation, how many properties do we end up seeing? I.e. could we call once to dump all of them, then pick things out in the JS layer?

File: vnext/Scripts/msbuildeval.ps1 (hunk 4 above)
Comment (dannyvv) [reply vnext/Scripts/msbuildeval.ps1:44]: .sln files are a very archaic file format. There have been many attempts to remove it but it never happend. Since .sln files have very strict control over configuration, i.e it can override that project Abc will build it's x86 when the solutoin is build using arm64 configuration. It can configure to skip building certain projects in certain configurations etc etc..

For MsBuild to handle this it actually generates an msbuild file under the covers that implements that special logic. This is the metaproj file. If you run a build using `/bl` you can actually see the engine building that metaproj and it is driving the build of the projects inthe .sln file.

File: vnext/Scripts/msbuildeval.ps1 [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    # Evaluate all of the Solution* properties and save into a collection
+    $globalProps = New-Object 'System.Collections.Generic.Dictionary[String,String]'
+    $solution.Properties | ForEach-Object -Process {
+        if ($_.Name.StartsWith("Solution")) {
+            $globalProps.Add($_.Name, $_.EvaluatedValue)
+        }
+    }
+
+    # Process the project file (with the Solution* properties we calculated before)
+    $project = [Microsoft.Build.Evaluation.Project]::new("$ProjectPath", $globalProps, "Current", $projectCollection)
```
Comment (dannyvv) [inline vnext/Scripts/msbuildeval.ps1:59]: note that the global props are technically per-project (especially for platform and configuration)

Comment (dannyvv) [reply vnext/Scripts/msbuildeval.ps1:59]: Also we should consider the environment variables as they control msbuild evaluation as well... I don't recall if the evaluatoin project required explicitly setting these as part of the projectcollection or not...

File: vnext/Scripts/msbuildeval.ps1 [hunk 8]
```diff
@@ -0,0 +1,73 @@
+# Evaluate an MSBuild property name
+param(
+    [Parameter(Mandatory=$true)]
```
Comment (dannyvv) [inline vnext/Scripts/msbuildeval.ps1:3]: I know that for react-native-windows it is not the case, but large msbuild based codebases do not have .sln files. They have directory traversal files and use custom private tooling to generate .sln files to edit in visual studio and limmit the number of projects. So technically for rnw projects that are part of larger builds this tool would not work...

Again, not a big deal, as our rnw msbuild files are not suited for integration in larger builds at the moment, but just sharing as FYI.

File: vnext/Scripts/msbuildeval.ps1 (hunk 6 above)
Comment (dannyvv) [reply vnext/Scripts/msbuildeval.ps1:67]: I think Julio was hinting at this too.

File: vnext/Scripts/msbuildeval.ps1 (hunk 3 above)
Comment (jonthysell) [reply vnext/Scripts/msbuildeval.ps1:47]: Because if you call msbuild on just a project file, then we don't get any solution variables like `SolutionFile`, which means any evaluations that use those values will be wrong.

File: vnext/Scripts/msbuildeval.ps1 (hunk 6 above)
Comment (jonthysell) [reply vnext/Scripts/msbuildeval.ps1:67]: Yeah, I just wanted to start with something "self-contained" since the only immediate requirement is being able to enable autolinking to determine which version of WinUI to use (and potentially fix #8591).

File: vnext/Scripts/msbuildeval.ps1 (hunk 7 above)
Comment (jonthysell) [reply vnext/Scripts/msbuildeval.ps1:59]: I'm not sure. Even my processing of the solution file to extract the Solution* variables is kind of problematic, since there may be other variables I'm missing.

File: vnext/Scripts/msbuildeval.ps1 (hunk 2 above)
Comment (jonthysell) [reply vnext/Scripts/msbuildeval.ps1:70]: Then you have to specify full paths, but yes that's an option.

File: vnext/Scripts/msbuildeval.ps1 (hunk 5 above)
Comment (jonthysell) [reply vnext/Scripts/msbuildeval.ps1:18]: Yeah, this logic isn't great, but I wanted to start with a "standalone" PS solution, and our CLI logic for MSBuild is all buried in JS.

File: vnext/Scripts/msbuildeval.ps1 (hunk 8 above)
Comment (jonthysell) [reply vnext/Scripts/msbuildeval.ps1:3]: We already have explicit checks elsewhere to make sure that a solution is present or we don't build and error out.



> > * Maybe this can be achieved just wrapping around `MSBuild.exe`?
> >   See `Vars` target at https://github.com/microsoft/react-native-windows/blob/main/vnext/Directory.Build.targets#L8
> 
> You have a cool idea here: We could have a target: `<Target Name="ExtractState">` that would write out a json file that we can then parse from js-side.

Yeah, we absolutely could, but then the question is can it be configured to output *all* variables, or do we have to hardcode the variables we want? That may be a faster, more scoped down option for our current use cases.

**Comment by jonthysell** [comment]
@NickGerleman , @dannyvv , @JunielKatarn 

Tagged as `needs dev design`, since this particular solution doesn't necessarily meet our needs (given the perf impact, and the limitations and assumptions I've made with dealing with MSBuild versions and solution files).

But, even if we go in the targets direction, be sure I'm saving this script somewhere in my PATH as a quick and dirty debugging tool. :)

**Comment by NickGerleman** [comment]
Bulk marking PRs without activity in the last 14 days as needing author feedback. This helps keep all active PRs visible on the first page of PRs. msftbot will close this PR in two weeks if the label is not removed. Please feel free to remove the label if you are still actively working on this.

**Comment by jonthysell** [comment]
I've updated the script to accept multiple properties.

**Comment by jonthysell** [comment]
I'm cloising this because I want to integrate the script into a PR that actually uses it.
